### ISSUE TITLE
[Proposal] artisan schedule:run --wait/block

### ISSUE THREAD

**Issue by DSpeichert** [laravel/ideas#1495]
This proposal has nothing to do with blocking while scheduler tasks are running (which already happens unless they are running in the background).

The feature would be useful when `artisan schedule:run` is ran as a docker container with `restart: always` policy, or any other process management tool that is not cron.

Currently it's possible to simply execute: `artisan schedule:run; sleep 60` but that will sometimes wait too long (if tasks are actually running, it will be `task_run_time+60`, instead of every 60 seconds.

This switch would avoid this unnecessary delay AND at the same time guarantee that `schedule:run` won't run every split seconds in a constant loop generating unnecessary log.

## What the `--wait/block` flag should do?
If there are no tasks to run currently, (`No scheduled commands are ready to run.`), block the process until the time of next task then exit cleanly. The `schedule:run` will be restarted and the task will actually run. Then next invocation will probably block again, or not - if there is a task congestion, they will all run very quickly after one another due to `restart: always` policy.

Note for systemd: this can also be useful with `Type=oneshot` and `Restart=always` policy there.

TL;DR: Overall, it saves burning CPU cycles while also enabling more on-time task execution.

**Comment by Namoshek** [comment]
I'm already using something like this. I simply created my own version of `schedule:run` based on the framework command:

```php
class ScheduleRun extends \Illuminate\Console\Scheduling\ScheduleRunCommand
{
    /**
     * The name and signature of the console command.
     *
     * @var string
     */
    protected $signature = 'schedule:run {--periodic : Run the schedule periodically}
                                         {--interval=60 : Seconds to wait between executions of the schedule}';

    /**
     * Execute the console command.
     *
     * @return void
     */
    public function handle()
    {
        if (!$this->option('periodic')) {
            $this->info('Running the schedule once.', OutputInterface::VERBOSITY_DEBUG);
            parent::handle();
            return;
        }

        $this->info('Running the schedule periodically.', OutputInterface::VERBOSITY_DEBUG);
        $interval = ((int) $this->option('interval')) * 1000000; // We use microseconds for better accuracy

        while (true) {
            $startOfExecution = now();
            parent::handle();
            $executionDuration = $this->diffInMicroseconds($startOfExecution, now());

            if ($executionDuration < $interval) {
                $sleep = $interval - $executionDuration;

                $sleepInSeconds = $sleep / 1000000;
                $this->info("Waiting for the next schedule execution in {$sleepInSeconds} seconds.", OutputInterface::VERBOSITY_DEBUG);

                usleep($sleep);
            }
        }
    }

    private function diffInMicroseconds(Carbon $date1, Carbon $date2)
    {
        $diff = $date1->diff($date2);

        return $diff->f * 1000000 // Microseconds as fraction of a second (float)
            + $diff->s * 1000000
            + $diff->i * 1000000 * 60
            + $diff->h * 1000000 * 60 * 60
            + $diff->days * 1000000 * 60 * 60 * 24; // Total number of days
    }
}
```

This is a working example of what I'm using (with the microseconds helper added to the class for simplicity).

The fact that I'm using this already speaks for itself I guess (highly welcomed in the core!). There might be some possible improvements to my implementation like signal handling or sleep drift prevention though (the calculated sleep times and sleeps themselves are by design not perfectly accurate, meaning we might need to correct a drift once in a while).
